File: bindings/lua/Polycode/Scene.lua
require "Polycode/EventDispatcher"

class "Scene" (EventDispatcher)

Scene.SCENE_3D = 0
Scene.SCENE_2D = 1
Scene.SCENE_2D_TOPLEFT = 2

function Scene:__getvar(name)
	if name == "clearColor" then
		local retVal = Polycode.Scene_get_clearColor(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Color"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "useClearColor" then
		return Polycode.Scene_get_useClearColor(self.__ptr)
	elseif name == "useClearDepth" then
		return Polycode.Scene_get_useClearDepth(self.__ptr)
	elseif name == "ambientColor" then
		local retVal = Polycode.Scene_get_ambientColor(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Color"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "fogColor" then
		local retVal = Polycode.Scene_get_fogColor(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Color"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "enabled" then
		return Polycode.Scene_get_enabled(self.__ptr)
	elseif name == "ownsChildren" then
		return Polycode.Scene_get_ownsChildren(self.__ptr)
	elseif name == "rootEntity" then
		local retVal = Polycode.Scene_get_rootEntity(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Entity"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "sceneMouseRect" then
		local retVal = Polycode.Scene_get_sceneMouseRect(self.__ptr)
		if retVal == nil then return nil end
		local __c = _G["Rectangle"]("__skip_ptr__")
		__c.__ptr = retVal
		return __c
	elseif name == "remapMouse" then
		return Polycode.Scene_get_remapMouse(self.__ptr)
	elseif name == "constrainPickingToViewport" then
		return Polycode.Scene_get_constrainPickingToViewport(self.__ptr)
	end
	if EventDispatcher["__getvar"] ~= nil then
		return EventDispatcher.__getvar(self, name)
	end
end

function Scene:__setvar(name,value)
	if name == "clearColor" then
		Polycode.Scene_set_clearColor(self.__ptr, value.__ptr)
		return true
	elseif name == "useClearColor" then
		Polycode.Scene_set_useClearColor(self.__ptr, value)
		return true
	elseif name == "useClearDepth" then
		Polycode.Scene_set_useClearDepth(self.__ptr, value)
		return true
	elseif name == "ambientColor" then
		Polycode.Scene_set_ambientColor(self.__ptr, value.__ptr)
		return true
	elseif name == "fogColor" then
		Polycode.Scene_set_fogColor(self.__ptr, value.__ptr)
		return true
	elseif name == "enabled" then
		Polycode.Scene_set_enabled(self.__ptr, value)
		return true
	elseif name == "ownsChildren" then
		Polycode.Scene_set_ownsChildren(self.__ptr, value)
		return true
	elseif name == "rootEntity" then
		Polycode.Scene_set_rootEntity(self.__ptr, value.__ptr)
		return true
	elseif name == "sceneMouseRect" then
		Polycode.Scene_set_sceneMouseRect(self.__ptr, value.__ptr)
		return true
	elseif name == "remapMouse" then
		Polycode.Scene_set_remapMouse(self.__ptr, value)
		return true
	elseif name == "constrainPickingToViewport" then
		Polycode.Scene_set_constrainPickingToViewport(self.__ptr, value)
		return true
	end
	if EventDispatcher["__setvar"] ~= nil then
		return EventDispatcher.__setvar(self, name, value)
	else
		return false
	end
end
function Scene:Scene(...)
	local arg = {...}
	if type(arg[1]) == "table" and count(arg) == 1 then
		if ""..arg[1].__classname == "EventDispatcher" then
			self.__ptr = arg[1].__ptr
			return
		end
	end
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.Scene(unpack(arg))
	end
end

function Scene:enableFog(enable)
	local retVal = Polycode.Scene_enableFog(self.__ptr, enable)
end

function Scene:setFogProperties(fogMode, color, density, startDepth, endDepth)
	local retVal = Polycode.Scene_setFogProperties(self.__ptr, fogMode, color.__ptr, density, startDepth, endDepth)
end

function Scene:setSceneType(newType)
	local retVal = Polycode.Scene_setSceneType(self.__ptr, newType)
end

function Scene:fixedUpdate()
	local retVal =  Polycode.Scene_fixedUpdate(self.__ptr)
end

function Scene:Update()
	local retVal =  Polycode.Scene_Update(self.__ptr)
end

function Scene:isEnabled()
	local retVal =  Polycode.Scene_isEnabled(self.__ptr)
	return retVal
end

function Scene:setEnabled(enabled)
	local retVal = Polycode.Scene_setEnabled(self.__ptr, enabled)
end

function Scene:setOverrideMaterial(material)
	local retVal = Polycode.Scene_setOverrideMaterial(self.__ptr, material.__ptr)
end

function Scene:getNumLights()
	local retVal =  Polycode.Scene_getNumLights(self.__ptr)
	return retVal
end

function Scene:doVisibilityChecking(val)
	local retVal = Polycode.Scene_doVisibilityChecking(self.__ptr, val)
end

function Scene:doesVisibilityChecking()
	local retVal =  Polycode.Scene_doesVisibilityChecking(self.__ptr)
	return retVal
end

function Scene:__delete()
	if self then Polycode.delete_Scene(self.__ptr) end
end

File: bindings/javascript/Polycode/Scene.js
require('Polycode/EventDispatcher')

function Scene() {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.Scene()
	}
	Object.defineProperties(this, {
		'clearColor': { enumerable: true, configurable: true, get: Scene.prototype.__get_clearColor, set: Scene.prototype.__set_clearColor},
		'useClearColor': { enumerable: true, configurable: true, get: Scene.prototype.__get_useClearColor, set: Scene.prototype.__set_useClearColor},
		'useClearDepth': { enumerable: true, configurable: true, get: Scene.prototype.__get_useClearDepth, set: Scene.prototype.__set_useClearDepth},
		'ambientColor': { enumerable: true, configurable: true, get: Scene.prototype.__get_ambientColor, set: Scene.prototype.__set_ambientColor},
		'fogColor': { enumerable: true, configurable: true, get: Scene.prototype.__get_fogColor, set: Scene.prototype.__set_fogColor},
		'enabled': { enumerable: true, configurable: true, get: Scene.prototype.__get_enabled, set: Scene.prototype.__set_enabled},
		'ownsChildren': { enumerable: true, configurable: true, get: Scene.prototype.__get_ownsChildren, set: Scene.prototype.__set_ownsChildren},
		'rootEntity': { enumerable: true, configurable: true, get: Scene.prototype.__get_rootEntity, set: Scene.prototype.__set_rootEntity},
		'sceneMouseRect': { enumerable: true, configurable: true, get: Scene.prototype.__get_sceneMouseRect, set: Scene.prototype.__set_sceneMouseRect},
		'remapMouse': { enumerable: true, configurable: true, get: Scene.prototype.__get_remapMouse, set: Scene.prototype.__set_remapMouse},
		'constrainPickingToViewport': { enumerable: true, configurable: true, get: Scene.prototype.__get_constrainPickingToViewport, set: Scene.prototype.__set_constrainPickingToViewport}
	})
}

Scene.SCENE_3D = 0
Scene.SCENE_2D = 1
Scene.SCENE_2D_TOPLEFT = 2

Scene.prototype = Object.create(EventDispatcher.prototype)

Scene.prototype.__get_clearColor = function() {
	var retVal = new Color("__skip_ptr__")
	retVal.__ptr = 	Polycode.Scene__get_clearColor(this.__ptr)
	return retVal
}

Scene.prototype.__set_clearColor = function(val) {
	Polycode.Scene__set_clearColor(this.__ptr, val.__ptr)
}

Scene.prototype.__get_useClearColor = function() {
	return Polycode.Scene__get_useClearColor(this.__ptr)
}

Scene.prototype.__set_useClearColor = function(val) {
	Polycode.Scene__set_useClearColor(this.__ptr, val)
}

Scene.prototype.__get_useClearDepth = function() {
	return Polycode.Scene__get_useClearDepth(this.__ptr)
}

Scene.prototype.__set_useClearDepth = function(val) {
	Polycode.Scene__set_useClearDepth(this.__ptr, val)
}

Scene.prototype.__get_ambientColor = function() {
	var retVal = new Color("__skip_ptr__")
	retVal.__ptr = 	Polycode.Scene__get_ambientColor(this.__ptr)
	return retVal
}

Scene.prototype.__set_ambientColor = function(val) {
	Polycode.Scene__set_ambientColor(this.__ptr, val.__ptr)
}

Scene.prototype.__get_fogColor = function() {
	var retVal = new Color("__skip_ptr__")
	retVal.__ptr = 	Polycode.Scene__get_fogColor(this.__ptr)
	return retVal
}

Scene.prototype.__set_fogColor = function(val) {
	Polycode.Scene__set_fogColor(this.__ptr, val.__ptr)
}

Scene.prototype.__get_enabled = function() {
	return Polycode.Scene__get_enabled(this.__ptr)
}

Scene.prototype.__set_enabled = function(val) {
	Polycode.Scene__set_enabled(this.__ptr, val)
}

Scene.prototype.__get_ownsChildren = function() {
	return Polycode.Scene__get_ownsChildren(this.__ptr)
}

Scene.prototype.__set_ownsChildren = function(val) {
	Polycode.Scene__set_ownsChildren(this.__ptr, val)
}

Scene.prototype.__get_rootEntity = function() {
	var retVal = new Entity("__skip_ptr__")
	retVal.__ptr = 	Polycode.Scene__get_rootEntity(this.__ptr)
	return retVal
}

Scene.prototype.__set_rootEntity = function(val) {
	Polycode.Scene__set_rootEntity(this.__ptr, val.__ptr)
}

Scene.prototype.__get_sceneMouseRect = function() {
	var retVal = new Rectangle("__skip_ptr__")
	retVal.__ptr = 	Polycode.Scene__get_sceneMouseRect(this.__ptr)
	return retVal
}

Scene.prototype.__set_sceneMouseRect = function(val) {
	Polycode.Scene__set_sceneMouseRect(this.__ptr, val.__ptr)
}

Scene.prototype.__get_remapMouse = function() {
	return Polycode.Scene__get_remapMouse(this.__ptr)
}

Scene.prototype.__set_remapMouse = function(val) {
	Polycode.Scene__set_remapMouse(this.__ptr, val)
}

Scene.prototype.__get_constrainPickingToViewport = function() {
	return Polycode.Scene__get_constrainPickingToViewport(this.__ptr)
}

Scene.prototype.__set_constrainPickingToViewport = function(val) {
	Polycode.Scene__set_constrainPickingToViewport(this.__ptr, val)
}


Scene.prototype.enableFog = function(enable) {
	Polycode.Scene_enableFog(this.__ptr, enable)
}

Scene.prototype.setFogProperties = function(fogMode,color,density,startDepth,endDepth) {
	Polycode.Scene_setFogProperties(this.__ptr, fogMode, color, density, startDepth, endDepth)
}

Scene.prototype.setSceneType = function(newType) {
	Polycode.Scene_setSceneType(this.__ptr, newType)
}

Scene.prototype.fixedUpdate = function() {
	Polycode.Scene_fixedUpdate(this.__ptr)
}

Scene.prototype.Update = function(elapsed) {
	Polycode.Scene_Update(this.__ptr, elapsed)
}

Scene.prototype.isEnabled = function() {
	return Polycode.Scene_isEnabled(this.__ptr)
}

Scene.prototype.setEnabled = function(enabled) {
	Polycode.Scene_setEnabled(this.__ptr, enabled)
}

Scene.prototype.setOverrideMaterial = function(material) {
	Polycode.Scene_setOverrideMaterial(this.__ptr, material)
}

Scene.prototype.getNumLights = function() {
	return Polycode.Scene_getNumLights(this.__ptr)
}

Scene.prototype.doVisibilityChecking = function(val) {
	Polycode.Scene_doVisibilityChecking(this.__ptr, val)
}

Scene.prototype.doesVisibilityChecking = function() {
	return Polycode.Scene_doesVisibilityChecking(this.__ptr)
}
